text/0000-target-feature.md

```diff
@@ -0,0 +1,551 @@
+
+Calling a function annotated with `#[target_feature]` on a host that does not
+support the feature invokes undefined behavior in LLVM, the assembler, and
+possibly the
+CPU.
+[See this comment](https://github.com/rust-lang/rfcs/pull/2045#issuecomment-311325202).
+
+## Relax inlining restrictions of `#[target_feature]`
+
+The rules proposed in this RFC for the interaction between `#[target_feature]`,
+inlining, and `#[always_inline]` are modeled after the rules in LLVM and are
+consistent with clang, e.g., see
+this [nice clang compilation error](https://godbolt.org/g/3ELXyJ)
+
+Rust, the language, could relax these rules. Whether doing so would require
+changes to LLVM hasn't been explored here.
+
+In particular, we could relax the rules to allow the "Example 1 (inlining)" to
+compile by using target feature boundaries within which code can be inlined, but
+across which code cannot be reordered:
+
+```rust
+#[target_feature="+sse3"]
+// -- sse3 boundary start (applies to fn arguments as well)
+fn baz() {
+  if std::cpuid::has_feature("avx") {
+    // -- sse3 boundary ends
+    // -- avx boundary starts
+    foo(); // might or might not be inlined, up to the inliner
+    // -- avx boundary ends
+    // -- avx boundary starts (the end and start of the boundary could be merged)
+    // bar is inlined here, its code cannot be reordered across the avx boudnary 
+    // -- avx boundary ends
+    // -- sse3 boundary starts
+    }
+  } 
+}
+// -- sse3 boundary ends (applies to drop as well)
+```
+
+## No segfaults from `unsafe` code
+
+Calling a `#[target_feature]`-annotated function on a platform that does not
+support it invokes undefined behavior. A friendly compiler could use run-time
+feature detection to check whether calling the function is safe and emit a nice
+`panic!` message. 
+
+This can be done, for example, by desugaring this: 
+
+```rust
+#[target_feature = "+avx"] unsafe fn foo();
+```
+
+into this:
+
+```rust
+#[target_feature = "+avx"] unsafe fn foo_impl() { ...foo tokens... };
+
+// this function will be called if avx is not available:
+unsafe fn foo_fallback() { 
+    panic!("calling foo() requires a target with avx support")
+}
+
+// run-time feature detection on initialization
+static foo_ptr: fn() -> () = if std::cpuid::has_feature("avx") { 
+    unsafe { foo_impl} }
+} else {
+    foo_fallback
+};
+
+// dispatches foo via function pointer to produce nice diagnostic
+unsafe fn foo() { foo_ptr() }
+```
+
+Since this is not required for safety, and can probably be done without an RFC,
+it is not proposed nor discussed further in this RFC.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Do we need to include run-time feature detection in std/rustc? Do we need to
+  stabilize run-time feature detection together with this RFC? What is the best
+  API for run-time feature detection?
+
+- Can we lift the restriction on `target_feature="-feature"` to provide
+  substractive features?
```

Comment:
Currently, the RFC only allows `#[target_feature = "+..."]`, that is, adding features over the default feature. This is an attempt to avoid non-sensical code like [this example that you showed](https://internals.rust-lang.org/t/pre-rfc-stabilization-of-target-feature/5176/26). 

Since I haven't given `#[target_feature = "-..."]` much thought I've listed it as an unresolved question in case somebody wants to give it a try.